### ISSUE TITLE
Failed preemptive refresh due to a conflict: committed value on key

### ISSUE THREAD

**Issue by JanCizmar** [cockroachdb/cockroach#94741]
**Describe the problem**

I am getting this error: 

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to a conflict: committed value on key /Table/123/8/2/0): "sql txn" meta={id=dcf48ead key=/Table/123/8/3/0 pri=0.03551682 epo=0 ts=1672851243.656743278,1 min=1672851232.686813476,0 seq=1012} lock=true stat=PENDING rts=1672851232.686813476,0 wto=false gul=1672851233.186813476,0
```

I don't think I am running it in multiple transactions, so it's weird.

**To Reproduce**

What did you do? Describe in your own words.

I am running some queries creating some rows. Right before commit I am running a query checking whether user doesn't exceeded their subscription limits. When I run this relatively complex query:

```
select
   (select sum(keyCount * languageCount) as translationCount
    from (select p.id as projectId, count(l.id) as languageCount
          from project as p
                   join language as l on l.project_id = p.id
          where p.organization_owner_id = ?
          group by p.id) as languageCounts
             join (select p.id as projectId, count(k.id) as keyCount
                   from project as p
                            join key as k on k.project_id = p.id
                   where p.organization_owner_id = ?
                   group by p.id) as keyCounts on keyCounts.projectId = languageCounts.projectId)
```

It fails on the described error on transaction commit. I also tried to run the query in separate transaction (which works fine in postgres), but in that case in just hangs forever in some kind of deadlock.

I tried to create a minimal reproducible example, but it fails really only in the few test cases, which unfortunately are part of private code. 

**Environment:**
 - CockroachDB version 22.2.1
 - Server OS: Linux/Mac
 - Client app JDBC

It might be related to this:
https://github.com/cockroachdb/cockroach/issues/52366

Sorry I didn't provide more info. Maybe someone with similar issue will provide more specific reproduce steps.

Jira issue: CRDB-23111

**Comment by jordanlewis** [comment]
@JanCizmar, please take a look at this documentation about retry errors in CockroachDB and see if it helps: https://www.cockroachlabs.com/docs/v22.2/transactions#client-side-intervention

TL;DR you need to handle serializable retry errors on the client side by retrying a transaction if you see that error. I'll close this as I don't think it's a bug. Also please feel free to join our community Slack or Forum: https://www.cockroachlabs.com/join-community/

**Comment by JanCizmar** [comment]
Hey, the odd thing is that I don't think it runs in multiple transactions. But I am not 100% about this, since the queries are executed by Hibernate, so it gets a bit confusing. Thanks anyway.

**Comment by ryb73** [comment]
Hi @JanCizmar, did you ever figure this out? I'm running into a similar problem.

**Comment by JanCizmar** [comment]
Hey! I think the problem was that my code wasn't ready for [isolation level `Serializable`](https://www.postgresql.org/docs/current/transaction-iso.html). I was basically trying to read the data, after I was changing it in the same transaction, which was fine with "read committed" (default in Postgres).
